File: macos/QMK Toolbox/USB/USBDevice.m
#import <IOKit/usb/IOUSBLib.h>

#import "USBDevice.h"

@implementation USBDevice

@synthesize service;

@synthesize manufacturerString;

@synthesize productString;

@synthesize vendorID;

@synthesize productID;

@synthesize revisionBCD;

- (id)initWithService:(io_service_t)service {
    if (self = [super init]) {
        self.service = service;

        self.manufacturerString = [self stringProperty:CFSTR(kUSBVendorString) forService:self.service];
        self.productString = [self stringProperty:CFSTR(kUSBProductString) forService:self.service];
        self.vendorID = [self vendorIDForService:self.service];
        self.productID = [self productIDForService:self.service];
        self.revisionBCD = [self ushortProperty:CFSTR(kUSBDeviceReleaseNumber) forService:self.service];
    }
    return self;
}

- (NSString *)stringProperty:(CFStringRef)property forService:(io_service_t)service {
    CFStringRef cfProperty = IORegistryEntryCreateCFProperty(service, property, kCFAllocatorDefault, kNilOptions);
    if (cfProperty != nil) {
        NSString *nsProperty = (__bridge NSString *)(cfProperty);
        CFRelease(cfProperty);
        return nsProperty;
    }
    return nil;
}

- (ushort)ushortProperty:(CFStringRef)property forService:(io_service_t)service {
    CFNumberRef cfProperty = IORegistryEntrySearchCFProperty(service, kIOServicePlane, property, kCFAllocatorDefault, kIORegistryIterateParents | kIORegistryIterateRecursively);
    if (cfProperty != nil) {
        NSNumber *nsProperty = (__bridge NSNumber *)(cfProperty);
        CFRelease(cfProperty);
        return [nsProperty unsignedShortValue];
    }
    return 0;
}

- (ushort)vendorIDForService:(io_service_t)service {
    return [self ushortProperty:CFSTR(kUSBVendorID) forService:service];
}

- (ushort)productIDForService:(io_service_t)service {
    return [self ushortProperty:CFSTR(kUSBProductID) forService:service];
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ %@ (%04X:%04X:%04X)", self.manufacturerString, self.productString, self.vendorID, self.productID, self.revisionBCD];
}

@end

File: macos/QMK Toolbox/USB/USBDevice.swift
import Foundation
import IOKit.usb

protocol USBDeviceProtocol {
    var service: io_service_t { get }

    var manufacturer: String? { get }
    var product: String? { get }

    var vendorID: UInt16 { get }
    var productID: UInt16 { get }
    var revisionBCD: UInt16 { get }
}

class USBDevice: USBDeviceProtocol, CustomStringConvertible {
    let service: io_service_t

    var manufacturer: String? {
        stringProperty(kUSBVendorString, service: service)
    }
    var product: String? {
        stringProperty(kUSBProductString, service: service)
    }

    var vendorID: UInt16 {
        vendorID(service: service)
    }
    var productID: UInt16 {
        productID(service: service)
    }
    var revisionBCD: UInt16 {
        ushortProperty(kUSBDeviceReleaseNumber, service: service)
    }

    init(service: io_service_t) {
        self.service = service
    }

    func stringProperty(_ property: String, service: io_service_t) -> String? {
        let cfProperty = IORegistryEntrySearchCFProperty(service, kIOServicePlane, property as CFString, kCFAllocatorDefault, IOOptionBits(kIORegistryIterateParents | kIORegistryIterateRecursively))
        return (cfProperty as? NSString) as? String
    }

    func ushortProperty(_ property: String, service: io_service_t) -> UInt16 {
        let cfProperty = IORegistryEntrySearchCFProperty(service, kIOServicePlane, property as CFString, kCFAllocatorDefault, IOOptionBits(kIORegistryIterateParents | kIORegistryIterateRecursively))
        return (cfProperty as? NSNumber)?.uint16Value ?? 0
    }

    func vendorID(service: io_service_t) -> UInt16 {
        ushortProperty(kUSBVendorID, service: service)
    }

    func productID(service: io_service_t) -> UInt16 {
        ushortProperty(kUSBProductID, service: service)
    }

    var description: String {
        String(format: "%@ %@ (%04X:%04X:%04X)", manufacturer ?? "", product ?? "", vendorID, productID, revisionBCD)
    }
}
